### PR TITLE
feat(phase4): dLLM grammar frontier bitmask + worker remask budget (#9 #10)

### PR DESCRIPTION
**Pull request by AlonKellner-RedHat**
## Summary

Phase 4 implementation for [issue #9](https://github.com/vllm-project/dllm-plugin/issues/9) (scheduler grammar safety / fixed `DRAFT_SIZE` blocks) and [issue #10](https://github.com/vllm-project/dllm-plugin/issues/10) (`DllmWorker` one-block path, remask integration).

This branch also aligns **`dllm-plugin` with vLLM 0.20.x** on the v2 GPU model runner path (`DllmGPUModelRunner`) so CPU CI (`--extra vllm`) and GPU integration smoke stay green.

## Changes

- **`grammar_utils`**: valid-prefix extraction for `grammar_bitmask`, frontier flat row indices, CPU bitmask apply aligned with vLLM semantics.
- **`DllmRuntimeScheduler`**: `get_grammar_bitmask` uses `validate_tokens` prefixes; `schedule()` attaches precomputed `GrammarOutput` + frontier metadata on `SchedulerOutput`; draft-token updates preserve full blocks (no grammar truncation in `update_draft_token_ids`).
- **`DllmRuntimeWorker`**: applies frontier-row mask to block logits; passes `grammar_extra_transfer` into remasking.
- **`Llada2DefaultRemaskingPolicy`**: optional `grammar_extra_transfer` augments transfer budget.
- **`DllmGPUModelRunner` / two-phase path**: compatibility with **vLLM 0.20** — forward `execute_model` kwargs (e.g. `is_profile`), `prepare_inputs` with **`BatchExecutionDescriptor`** / padded **`InputBatch`** (`num_reqs_after_padding`, `seq_lens_cpu_upper_bound`), **`ExecuteModelState`** in `sample_tokens`, and **`AsyncOutput`** construction without removed `output_copy_event`.
- **`mock_llada2`**: `Attention` import order for current `vllm.model_executor.layers.attention` layout.
- **Deps / docs**: optional `vllm` pin and lock refresh where applicable; operator/contributing notes as in branch.
- **Tests**: `tests/conftest.py` (`vllm_dllm_plugin` alias), `tests/test_grammar_utils.py`, GPU-target tests unchanged in intent (mock stack + MRV2 + regex grammar).
- **Docs**: `docs/OPERATOR_LLaDA2.md` structured-output section.

## Validation

- **GitHub Actions**: `ci` (Python 3.10–3.13) and **`vllm-extra`** (3.12, `--extra vllm`) expected to pass with the updated pins and runner fixes.
- **GPU (maintainer / fork evidence)**: Helm chart `tools/helm/dllm-plugin-gpu-test` against fork branch **`feat/phase4-dllm-grammar-worker-issues-9-10`** completed **`JOB_COMPLETE`** (job instance `dllm-plugin-gpu-mrv2-20260430-120109`), exercising mock **`LLM.generate`**, **`DllmGPUModelRunner` monkeypatch**, **regex structured-output grammar**, and **two-phase** tests on **CUDA + vLLM 0.20.0**.

Scope of GPU proof is **mock-stack smoke** (not full production LLaDA2 weights or all grammar backends).

## vLLM companion patch (runtime)

For structured outputs end-to-end, this plugin still expects upstream **`SchedulerOutput` / `EngineCore`** plumbing for dLLM grammar metadata and draft handoff where not yet in a given vLLM revision—coordinate with the matching **vLLM** change set or pin documented in this repo.

Closes #9  
Closes #10

Made with [Cursor](https://cursor.com)

---

### Amendments (second-opinion review + simplification)

- **Phase:** 4 — **Closes #9**, **Closes #10** (orchestration: [#19](https://github.com/vllm-project/dllm-plugin/issues/19)).
- **HARD deps:** vLLM **0.20.x** (`pyproject` / `uv.lock`); companion engine / `SchedulerOutput` plumbing as documented in-repo.
- **CONTRIBUTING:** paths corrected to **`dllm_plugin/`** (not `vllm_dllm_plugin/`).
- **`DllmGPUModelRunner`:** **0.20-only** — removed legacy 0.14 paths, tuple `execute_model_state`, and optional `copy_event` handling; `prepare_inputs` lives on **`_GPUModelRunnerPrepareInputsFork`** with overridable **`get_expand_idx_mapping_block_size`** (dLLM widening only in `DllmGPUModelRunner`).
- **Docs:** two-stage grammar + test env `VLLM_DLLM_SKIP_FIRST_BLOCK_SEED` in **`docs/OPERATOR_LLaDA2.md`**; **`docs/DESIGN_MVP.md`** fork / hook note.

---

### Local second-opinion follow-ups (2026-04-30)

- **PyPI vs companion:** Operator guide clarifies pin ≠ guaranteed `dllm_*` / engine parity; GPU/Helm remains E2E signal (**#2**).
- **`take_dllm_draft_token_ids`:** Documented deviation from literal #10 runner hook naming (DESIGN / CONTRACTS / OPERATOR + docstrings).
- **`prepare_inputs` fork:** Rebase baseline called out (**v0.20.0** tree URL) in `_GPUModelRunnerPrepareInputsFork`.
- **`num_invalid_spec_tokens`:** Comment that dLLM path leaves the map empty unless upstream needs otherwise.
- **Async + dLLM:** Inline note in `sample_tokens` next to `AsyncOutput` return.
- **Tests:** `grammar.validate_tokens` stub coverage for `valid_prefix_tokens_for_draft` and SO scheduling prefix path (`tests/test_grammar_utils.py`).

### CODE REVIEW

**Comment by AlonKellner-RedHat** [comment]
**Companion vLLM patch (local workspace):** branch `feat/v1-dllm-engine-scheduler-output-hooks` touches `vllm/v1/core/sched/output.py` and `vllm/v1/engine/core.py`. Push it from a vLLM fork and open an upstream PR alongside this plugin PR—the runtime needs both.

File: dllm_plugin/runtime_worker.py [hunk 1]
```diff
@@ -240,88 +258,47 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
...
-        self._dllm_last_draft_token_ids = (
-            cast(Any, DraftTokenIds)(
-                req_ids=next_req_ids,
-                draft_token_ids=next_blocks,
-            )
-            if next_req_ids
-            else None
-        )
-        return output
+        return super().execute_model(scheduler_output)
```
Comment (AlonKellner-RedHat) [inline dllm_plugin/runtime_worker.py]: This does nothing now, can be safely removed
